crates/pallet-domains/src/domain_registry.rs

```diff
@@ -0,0 +1,149 @@
+    pub target_bundles_per_block: u32,
+}
+
+impl DomainConfig {
+    pub(crate) fn from_genesis<T: Config>(
+        genesis_domain: &GenesisDomain<T::AccountId>,
+        runtime_id: RuntimeId,
+    ) -> Self {
+        DomainConfig {
+            domain_name: genesis_domain.domain_name.clone(),
+            runtime_id,
+            max_block_size: genesis_domain.max_block_size,
+            max_block_weight: genesis_domain.max_block_weight,
+            bundle_slot_probability: genesis_domain.bundle_slot_probability,
+            target_bundles_per_block: genesis_domain.target_bundles_per_block,
+        }
+    }
+}
+
+#[derive(TypeInfo, Debug, Encode, Decode, Clone, PartialEq, Eq)]
+pub struct DomainObject<Number, Hash, AccountId> {
+    /// The address of the domain creator, used to validate updating the domain config.
+    pub owner_account_id: AccountId,
+    /// The consensus chain block number when the domain first instantiated.
+    pub created_at: Number,
+    /// The hash of the genesis execution receipt for this domain.
+    pub genesis_receipt_hash: Hash,
+    /// The domain genesis config.
+    pub domain_config: DomainConfig,
+}
+
+pub(crate) fn can_instantiate_domain<T: Config>(
+    owner_account_id: &T::AccountId,
+    domain_config: &DomainConfig,
+) -> Result<(), Error> {
+    ensure!(
+        domain_config.domain_name.len() as u32 <= T::MaxDomainBlockSize::get(),
+        Error::DomainNameTooLong,
+    );
+    ensure!(
+        RuntimeRegistry::<T>::contains_key(domain_config.runtime_id),
+        Error::RuntimeNotFound
+    );
+    ensure!(
+        domain_config.max_block_size <= T::MaxDomainBlockSize::get(),
+        Error::ExceedMaxDomainBlockSize
+    );
+    ensure!(
+        domain_config.max_block_weight.ref_time() <= T::MaxDomainBlockWeight::get().ref_time(),
+        Error::ExceedMaxDomainBlockWeight
+    );
+    ensure!(
+        domain_config.target_bundles_per_block != 0
+            && domain_config.target_bundles_per_block <= T::MaxBundlesPerBlock::get(),
+        Error::InvalidBundlesPerBlock
+    );
+
+    // `bundle_slot_probability` must be `> 0` and `≤ 1`
+    let (numerator, denominator) = domain_config.bundle_slot_probability;
+    ensure!(
+        numerator != 0 && denominator != 0 && numerator <= denominator,
+        Error::InvalidSlotProbability
+    );
+
+    ensure!(
+        T::Currency::free_balance(owner_account_id) >= T::DomainInstantiationDeposit::get(),
+        Error::InsufficientFund
+    );
+
+    Ok(())
+}
+
+pub(crate) fn do_instantiate_domain<T: Config>(
+    domain_config: DomainConfig,
+    owner_account_id: T::AccountId,
+    created_at: T::BlockNumber,
+) -> Result<DomainId, Error> {
+    let domain_obj = DomainObject {
+        owner_account_id: owner_account_id.clone(),
+        created_at,
+        // TODO: drive the `genesis_receipt_hash` from genesis config through host function
```

Comment:
I prefer to handle this in the block tree PR to initialize the genesis block of the block tree at the same time, also because the host function still needs some revisions in order to use.